### PR TITLE
Implement purls for npm packages

### PR DESCRIPTION
**Pull request by chmeliik**
STONEBLD-1112

```
- registry dependencies: pkg:npm/name@version
- git dependencies: pkg:npm/name[@version]?vcs_url=url[#subpath]
- http dependencies: pkg:npm/name[@version]?download_url=url
- file dependencies: treated as git dependencies from the source repo
- main package: treated as a `file:.` dependency
```

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
PTAL

depends on #194 

Unit tests TBD

**Comment by chmeliik** [comment]
Added a registry.yarnpkg.com handling as per https://github.com/containerbuildsystem/cachito/pull/862, good catch Taylor

I can update the the downloading code accordingly after it's merged

**Comment by lkolacek** [comment]
Looks great so far

**Comment by chmeliik** [comment]
Added unit test for v2 lockfile with workspaces at a subpath

I think that's all the unit tests I wanted to implement, so now all that's missing should be

- `?checksum=` for https deps
- github: gitlab: bitbucket: protocol handling

**Comment by chmeliik** [comment]
All done, PTAL :slightly_smiling_face: 

You might point out the duplicated logic between downloading vs. purl generation - it is mostly refactored away in the last commit

**Comment by taylormadore** [comment]
I've tested all of the cases I can think of and the output looks perfect so far :rocket: 
I'll finish a more detailed review in the morning

File: cachi2/core/package_managers/npm.py [hunk 1]
```diff
@@ -294,6 +289,28 @@ def get_purl(
...
+NormalizedUrl = NewType("NormalizedUrl", str)
+
+
+def _normalize_resolved_url(resolved_url: str) -> NormalizedUrl:
+    if resolved_url.startswith(("github:", "gitlab:", "bitbucket:")):
+        resolved_url = _update_vcs_url_with_full_hostname(resolved_url)
+    return NormalizedUrl(resolved_url)
+
+
+def _classify_resolved_url(
```
Comment (lkolacek) [inline cachi2/core/package_managers/npm.py:301]: :rocket:

**Review by lkolacek — APPROVED** [review]
Looks amazing, Thank you! :+1:

**Review by ejegrova — APPROVED** [review]
LGTM :+1:

**Pull request merged**